### DISCUSSION TITLE
is x-html not working properly in an x-for template?

### DISCUSSION THREAD

**Discussion by leobm** [alpinejs/alpine#2765] [7. Bugs (archived)] [answered]
I think there is a bug with x-html. I could isolate the problem now.
Apparently it doesn't work as expected when it is in an x-for template.
https://gist.github.com/leobm/177690300d2d879bf8906cb3c22eeb92

posA elements are displayed anyway or do I have the wrong expectations?

**My expectation is that the posA elements will not be displayed.**

**Comment by SimoTod** [comment, marked as answer]
the mutation observer doesn't run during x-for so dynamically inserted content won't be picked up by alpine (https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives/x-for.js#L188).

I'm not sure there is an easy way to fix it, maybe deferring x-html using queueMicroTask would work.

**Reply by leobm** [reply]
ok, thanks, i thought i was already stupid. 

**@kwoka had suggested to put a :key on the elements, but that didn't help much either.**

-------------------------------------------------------------------------------------------------------------------------
I got it fixed with Alpine.morph. But then I had other problems. 
There was my problem when I make several Alpine.morph in the page. Since this runs asynchronously they interfere with each other. I only managed it by running the Alpine.morph operation syncronously (I always wait until a promise morph is finished and then start the next morph).

**Reply by SimoTod** [reply]
I think this can be reasonably classified as a bug. Let me know if you want to work on a PR, otherwise I'll give it a go tonight or tomorrow

**Reply by leobm** [reply]
the x-for bug or Alpine.morph problem?

Edit:
here my Alpine.morph hack 
https://gist.github.com/leobm/1987f8d4748825a0e798f6d2efbd5bf6

without my hack https://gist.github.com/leobm/524fb1411684e027e8858700a77e5b2e
I see four times posD in the page

**Reply by SimoTod** [reply]
the x-for bug

**Reply by leobm** [reply]
ok, this one is maybe one step too high for me currently, i would have to understand the code of alpinejs a little bit more. ;)

**Reply by SimoTod** [reply]
I sent a Pr for the x-for bug.
The morph bug is caused by a code refactoring in 3.9.1 (some variables have been made global and because morph is async, 2 sequential calls will get mixed). Your example works correctly with 3.8.0

**Reply by leobm** [reply]
ok , cool, the fix looks really simple. What are the implications of the fix? Why was the init for cloned nodes previously in mutateDom?

the problem with morph and the global variables I had already thought, but I thought it had some reason why it was done that way. I think, I use morph eigentlch also for a wrong purpose. If I understand correctly it is rather used for things like Liveview or Livewire, where you have mostly then always only one morph in a page?

Thank you for taking care of the problem so quickly. It really helps me because I'm using Alpinejs for the first time in a professional product.
